python/raf/_tvm_op/reduce.py

```diff
@@ -61,31 +67,306 @@ def schedule_sum(attrs, outs, target):
+                const_size = 0
+
+            if vector_width > 1:
+                fused, vec = sch[out].split(fused, vector_width)
+                sch[out].vectorize(vec)
+
+            if need_block_split:
+                xo, xi = sch[out].split(fused, factor=num_thread * max_block)
+                bx, tx = sch[out].split(xi, factor=num_thread)
+                sch[out].reorder(bx, tx, xo)
+                sch[out].bind(bx, _tvm.te.thread_axis("blockIdx.x"))
+                sch[out].bind(tx, _tvm.te.thread_axis("threadIdx.x"))
+            else:
+                if const_size != 0 and const_size < num_thread:
+                    bx, tx = sch[out].split(fused, factor=const_size)
+                else:
+                    bx, tx = sch[out].split(fused, factor=num_thread)
+                sch[out].bind(tx, _tvm.te.thread_axis("threadIdx.x"))
+                sch[out].bind(bx, _tvm.te.thread_axis("blockIdx.x"))
+    return sch
+
+
 @schedule_sum.register(["cuda", "gpu"])
 def schedule_sum_cuda(attrs, outs, target):
     # pylint: disable=unused-argument
-    # pylint: disable=invalid-name
     def get_num_elements(axes):
         extents = [int(iv.dom.extent) for iv in axes]
         n_elems = 1
         for extent in extents:
             n_elems *= extent
         return n_elems
 
+    def get_sum_input(tensor):
+        operator = tensor.op
+        if len(operator.input_tensors) != 1:
+            return None
+
+        input_tensor = operator.input_tensors[0]
+        if isinstance(input_tensor.op, _tvm.te.PlaceholderOp):
+            return input_tensor
+        return None
+
     with target:
         out = outs[0]
         num_out_elements = get_num_elements(out.op.axis)
         num_reduce_elements = get_num_elements(out.op.reduce_axis)
 
-        # We want to saturate the GPU cores by parallelization. There are 2 scenarios
-        # 1) Reduce dimension is small - In this case, each thread is responsible for reduction.
-        # The parallelization is across the output elements.
-        # 2) Reduce dimension is large - We want to parallelize the reduction to keep the GPU busy.
-        # Here we fall back to TVM schedule.
+        # Whether the last axis is reduced. Note that axis=-1 should already be proceed in advance.
+        reduce_last_axis = False
+        input_tensor = get_sum_input(out)
+        if input_tensor is not None:
+            # Only try to analyze the workload with sum as the last op.
+            reduce_axis = [False for _ in range(len(input_tensor.shape))]
+            for axis in attrs.axis:
+                reduce_axis[axis.value] = True
+            if attrs.exclude == 1:
+                reduce_axis = [not axis for axis in reduce_axis]
+            reduce_last_axis = reduce_axis[-1]
+
+        # We attempt to saturate the GPU cores by parallelization, so we dispatch
+        # the sum workloads to two schedules based on their reduction length.
         if num_out_elements > num_reduce_elements:
```

Comment:
Just curious: is there any insight behind this condition? What is the benefit of using such a condition over choosing heuristic based on the absolute number of elements?

Reply:
Good question. Since we use different parallelism strategies (parallelize on output elements or reduction elements) for these cases, looks like parallelize the large one is always beneficial (at least you get more parallelism...)